src/driver/kvm.rs

```diff
@@ -474,6 +455,884 @@ mod tests {
+        let sregs: kvm_sregs = Default::default();
+        let msrs = KvmMsrs::new();
+        kvmi_mock
+            .expect_get_registers()
+            .times(1)
+            .return_once(move |_| Ok((regs, sregs, msrs)));
+        let kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        let vcpu: u16 = 0;
+        let result = kvm.read_registers(vcpu);
+
+        assert!(result.is_ok(), "Expected ok, got error instead!");
+    }
+
+    #[test]
+    fn test_read_registers_fails_if_kvmi_get_registers_fails() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock.expect_get_registers().times(1).returning(|_| {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "something went wrong",
+            ))
+        });
+        let kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+
+        let vcpu: u16 = 0;
+        let result = kvm.read_registers(vcpu);
+
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_cr_intercept_succeeds() {
+        let mut kvmi_mock = MockKVMi::default();
+        kvmi_mock
+            .expect_control_cr()
+            .with(eq(0), function(|x| matches!(x, KVMiCr::Cr3)), eq(true))
+            .times(1)
+            .returning(|_, _, _| Ok(()));
+        setup_kvmi_mock(&mut kvmi_mock);
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+
+        let vcpu: u16 = 0;
+        let intercept_type = InterceptType::Cr(CrType::Cr3);
+        let enabled: bool = true;
+        let result = kvm.toggle_intercept(vcpu, intercept_type, enabled);
+
+        assert!(result.is_ok(), "Expected ok, got error instead!");
+    }
+
+    #[test]
+    fn test_cr_intercept_fails_if_kvmi_control_cr_fails() {
+        let mut kvmi_mock = MockKVMi::default();
+        kvmi_mock
+            .expect_control_cr()
+            .times(1)
+            .with(eq(0), function(|x| matches!(x, KVMiCr::Cr3)), eq(true))
+            .times(1)
+            .returning(|_, _, _| {
+                Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "something went wrong",
+                ))
+            });
+        setup_kvmi_mock(&mut kvmi_mock);
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
```

Comment:
Something that came to my mind: This part is also almost always the same right? Perhaps you could write another function the yields a kvm object completely initialized and set up. If you have the time, I don't deem this necessary. But it would certainly reduce the amount of duplicate setup code alot.

Reply:
Ah, you know what? Don't bother with that. You have enough on your plate as it is.